src/Strategy/Python/Setuptools.hs

```diff
@@ -7,29 +7,41 @@ module Strategy.Python.Setuptools (
 ) where
 
 import App.Fossa.Analyze.Types (AnalyzeProject, analyzeProject)
-import Control.Carrier.Output.IO
 import Control.Effect.Diagnostics (Diagnostics, context)
 import Control.Effect.Diagnostics qualified as Diag
+import Control.Effect.Reader (Reader)
 import Data.Aeson (ToJSON)
 import Data.List (isInfixOf, isSuffixOf)
 import Data.Maybe (maybeToList)
 import Data.Text (Text)
-import Discovery.Walk
-import Effect.ReadFS
+import Discovery.Filters (AllFilters, withToolFilter)
+import Discovery.Walk (
+  WalkStep (WalkContinue),
+  fileName,
+  findFileNamed,
+  walkWithFilters',
+ )
+import Effect.ReadFS (Has, ReadFS)
 import GHC.Generics (Generic)
 import Graphing (Graphing)
-import Path
+import Path (Abs, Dir, File, Path)
 import Strategy.Python.ReqTxt qualified as ReqTxt
 import Strategy.Python.SetupPy qualified as SetupPy
-import Types
+import Types (
+  Dependency,
+  DependencyResults (..),
+  DiscoveredProject (..),
+  DiscoveredProjectType (SetuptoolsProjectType),
+  GraphBreadth (Partial),
+ )
 
-discover :: (Has ReadFS sig m, Has Diagnostics sig m) => Path Abs Dir -> m [DiscoveredProject SetuptoolsProject]
-discover dir = context "Setuptools" $ do
+discover :: (Has ReadFS sig m, Has Diagnostics sig m, Has (Reader AllFilters) sig m) => Path Abs Dir -> m [DiscoveredProject SetuptoolsProject]
+discover dir = withToolFilter SetuptoolsProjectType $ context "Setuptools" $ do
```

Comment:
I'm a bit hesitant about this - It couples concerns of filtering within individual discovery and findProject function. Although, I can't pin a better alternative at the moment (maybe some form of factory?). Let me have some time with this. 

These are some of the thoughts that came at the moment, although I need some time to refine my thoughts. 

1) why not move this (withToolFilter a $ ..) decorator to `App.Fossa.Analyze.Discover`?
2) since this ProjectType lives as a value (declared) in DiscoveredProject as opposed to type makes this bit awkward - otherwise, maybe with `-XTypeApplications` this can be easier to filter, and have more strict guarantees?

Reply:
So, the `DiscoveredProjectType` is not deterministic for a few `discover` functions, so we can't make a filter that's generic for all of them that's automatically applied.  Even if we could, and did: type-enforcing those constant values is _really hard_ without fully dependent types, which haskell does not really have.

Reply:
While this setup is not error-proof, it's incredibly unobtrusive, and easily human-verifiable.  It's also very unlikely to change.

Reply:
Additionally, I put these into their associated strategies modules _intentionally_.  Each strategy has to decide _for itself_ whether to use the filters or not.  In some cases, it doesn't make sense to do so, particularly where we cross-pollinate strategies (see node/yarn, and scala/maven).

Reply:
This PR should be seen as a band-aid that allows us to deliver the general discovery exclusion feature.  The _real_ problem is that we have a 2-phase system (discovery/analysis), when we need a 3-phase system:
1. Walk the tree and note _potential_ project locations
2. Iterate through valid locations from step 1 to identify valid project roots (currently we do this in discovery as part of `findProjects`)
3. analyze (no change here).

Reply:
If you can find a way to make this less error-prone without sacrificing conciseness, introducing a massive refactor or adding complex type-level trickery, I'm 100% on board.  But I already went down those rabbit holes and I found nothing that didn't make things worse.

Reply:
> If you can find a way to make this less error-prone without sacrificing conciseness, introducing a massive refactor or adding complex type-level trickery, I'm 100% on board. But I already went down those rabbit holes and I found nothing that didn't make things worse.

I'm slowly coming to the same conclusion without having to redo/refactor the entirety of discovery plumbing.

Reply:
Posterity: We discussed this offline and agreed that this solution is good enough for now, but it would definitely be nice to have a less error-prone version.

Also, the addition of `simpleDiscover` can hopefully help with this.